File: listas/l6/9.cpp
#include <bits/stdc++.h>

using namespace std;

int opt[300][300];
vector<int> x { 10, 1, 7, 7 };
vector<int> s { 8, 4, 2, 1 };

int solve(){
    int n = x.size();

    for(int j=0;j<n;j++)
        opt[n-1][j] = min(x[n-1], s[j]);

    for(int i=n-2;i>=0;i--){
        for(int j=0;j<=i;j++){
            opt[i][j] = max(opt[i+1][0], min(x[i], s[j]) + opt[i+1][j+1]);
        }
    }

    return opt[0][0];
}

int main(){

    cout << solve() << endl;

    return 0;
}

File: listas/l6/9.py
x = [10, 1, 7, 7]
s = [8, 4, 2, 1]
n = len(x)

def solve():
    # opt = [ min(x[n-1], s[j]) for j in range(n) ]
    opt = [[0 for i in range(n)] for j in range(n)]

    for j in range(n):
        opt[n-1][j] = min(x[n-1], s[j])

    for i in reversed(range(n-1)):
        for j in range(i+1):
            opt[i][j] = max(
                            opt[i+1][0], # nothing is processed on day i
                            min(x[i], s[j]) + opt[i+1][j+1] # process min(x[i], s[j])
                        )

    return opt[0][0]

print('Maximun number of terabytes processed is {}'.format(solve()))
